verida/vault-mobile
# [EPIC] Upgrade WalletConnect to v2

### 💭 What's changed?

- Upgraded to [`WalletConnect v2`](https://github.com/WalletConnect/web-examples/tree/main/wallets/react-wallet-v2)!
  - Removed all the old dependencies and replaced with the newer APIs.
  - Stopped storing state inside of Redux for `walletConnect`.
    - WalletConnect persists and manages all of the associated state with connected wallets, so we can treat that as the source of truth and ensure we don't get into instances of stale or invalid connection states.
  - Removed the V1/V2 flows and replaced with a unified WalletConnect flow.
    - I've tried to preserve the intention of both flows wherever possible, though some newer data models made some of the older components no longer available.
  - We no longer need to cache the relationships between WalletConnect and a Verida Wallet ourselves - instead, these are inferred dynamically based upon the WalletRequest `request` object. This reduces the number of places we need to interact with Redux when using WalletConnect to just a single function whose only responsibility is to provide a `VeridaWallet` for a given `request`.
  - I separated a lot of the old WalletConnect components into more components which are narrower in scope.
    - Aside from making things a little easier to read and buying better reuse, it is also a bit more efficient to do it this way. For example, when rendering lots of JSX in a loop, when moving that JSX inside of a dedicated component there's a lot less you need to evaluate and you can also take advantage of hooks to reduce the reallocation workload.
- It is now much easier to be multichain!
  - All of our config objects representing different chains have been reduced to a single source of truth that we refer to for WalletConnect, finding RPC URLs, looking up high-level metadata like names, etc.
  - Corresponding business logic for each chain protocol type is now enforced using an `enum` `SupportedCaipProtocolStandard`. If we want to add a new blockchain, just add a new `SupportedCaipProtocolStandard` and all of the necessary parts of the codebase you need to implement will report a compile-time exception to make you aware that they need attention!
  - All of our WalletConnect-related flows go through common handlers - this means that when adding a new chain, you only need to focus on the details of that specific chain; meanwhile the rest of the common WalletConnect architecture will remain consistent.
    - This way, when we add a new chain we still provide a consist and robust WalletConnect experience. 
- Added more robust handling for CAIP identifiers.
  - It was really difficult for me to understand based on a `string` whether we were dealing with a `chainId`, a full CAIP identifier, a blockchain type, etc. Here, we've added a new high-level type called the `ParsedCaipType`, which ensures we represent the source data faithfully with little subjective interpretation as to what it could represent.
    - The `ParsedCaipType` also supports generic typing, so we can narrow particular regions of the codebase (i.e. NEAR-specific) to only specific types of identifiers. 
    - Used the `ParsedCaipType` to filter the types of Wallets we permit users to select when making a connection.
      - Previously, we would erroneously recommend the user connect a NEAR wallet for an EVM chain, or a Polygon Mumbai wallet for Ethereum Goerli, for example. 
- Added a simple unit testing suite using [`jest`](https://github.com/jestjs/jest).
  - It is now possible to test and lock in the expectations of pure functions. This is really useful because we can test if our expectations of certain operations get regressed over time, and in a way it serves like living documentation.
  - Jest allows us to use test snapshots, which let us automatically generate blobs of validation data to test against.
    - This makes it really easy to add new tests.
  - This is now run on CI, ensuring CI will fail unless all tests are passing.
- Refactored NEAR and EIP155 implementations.
  - Using the `features/` directory, I tried to separate these implementations to respect their differences and prevent unnecessary coupling.
  - We refactored away from some of the WalletConnect designs in order to better suit Verida's internal business logic.
    - There's some further work needed here - it seems strange to me that WalletConnect would have every wallet application duplicate the same code from their Wallet examples; it would be really nice to move these to a generic library that we could import from. As it stands, whenever we add a new chain we have to do a lot of work that could be skipped entirely with a sufficiently abstract adapter pattern, and we could benefit from the findings and improvements contributed by other wallets.
- Continued to remove code which appeared to be unused. 

### 🧪 How to test these changes?

- Try connecting to different applications and performing different transactions.
- https://react-app.walletconnect.com/
- See if your old wallet selection seems consistent.
  - I don't know enough of the Verida App so I'm not sure if there are any specific flows or UX I might have regressed, but it is likely.

### 😨 Anything to be aware of?

- I have removed the `AsyncStorage` implementation for the NEAR keystore, we instead use the NEAR `VeridaWallet` that lives in Redux.
  - The reason I removed the `AsyncStorage` version of the NEAR keystore was because it didn't seem necessary if we were already creating `VeridaWallet`s through a common pattern, and it is my understanding that using `AsyncStorage` just came from the WalletConnect examples, but that doesn't mean we _have_ to do it this way.
  - I'm not sure if this means that the current wallets on device will no longer be available, though my impression was we were using an unrelated NEAR wallet to what was being exported from Redux.
  - If there are indeed two wallet instances for NEAR in the current wallet path, we can introduce a migration pattern if it's necessary to hold onto the old wallet.
- Although we support lots of different EVM chains, we still require dedicated Verida wallet instances for the different chains. It looks like at the moment we only create wallets for Polygon and Ethereum, so it would be good to expand the scope of this process to generate wallets for all EVM chains we support.
- I'm pretty sure the Ethereum transaction request window has changed substantially from what's currently distributed (it now looks the same as the NEAR transaction request window, since they all go through common components.)
  - I haven't looked at any of the designs yet. I know we have a dedicated task for this, so I'm hoping to catch this in a subsequent PR. 
- I needed to add a Polyfill to allow the `InMemoryKeystore` to play nicely with a `privateKey`.
- I needed to duplicate a source file which comes from `@near-wallet-selector/wallet-utils`. The metro bundler was simply not happy with the file I needed when pulled from the SDK. Hopefully this can be deleted later.
